src/render/integrator.cpp

```diff
@@ -619,23 +627,28 @@ AdjointIntegrator<Float, Spectrum>::render(Scene *scene,
             evaluate = true;
         }
 
-        size_t wavefront_size = (size_t) film_size.x() *
-                                (size_t) film_size.y() * (size_t) spp_per_pass,
-               wavefront_size_limit =
-                   dr::is_llvm_v<Float> ? 0xffffffffu : 0x40000000u;
-
-        if (wavefront_size > wavefront_size_limit)
-            Throw("Tried to perform a %s-based rendering with a total sample "
-                  "count of %zu, which exceeds 2^%zu = %zu (the upper limit "
-                  "for this backend). Please use fewer samples per pixel or "
-                  "render using multiple passes.",
-                  dr::is_llvm_v<Float> ? "LLVM JIT" : "OptiX",
-                  wavefront_size, dr::log2i(wavefront_size_limit + 1),
-                  wavefront_size_limit);
+        size_t wavefront_size_limit =
+            dr::is_llvm_v<Float> ? 0xffffffffu : 0x40000000u;
+
+        if (samples_per_pass > wavefront_size_limit) {
+            spp_per_pass /= (samples_per_pass + wavefront_size_limit - 1) /
+                            wavefront_size_limit;
+            n_passes = spp / spp_per_pass;
+            samples_per_pass = (size_t) film_size.x() * (size_t) film_size.y() *
+                               (size_t) spp_per_pass;
+
+            Log(Warn,
+                "The requested rendering task involves %zu Monte Carlo "
+                "samples, which exceeds the upper limit of 2^%zu = %zu for "
+                "this variant. Mitsuba will instead split the rendering task "
+                "%zu smaller passes to avoid exceeding the limits.",
+                samples_per_pass, dr::log2i(wavefront_size_limit + 1),
+                wavefront_size_limit, n_passes);
+        }
```

Comment:
It might be worth abstracting this logic into a separate function and adding bindings as we will want to use this in the Python integrators as well.

Reply:
Hi Sebastien -- the Python side is likely going to be more complicated (we do not support the multi-pass rendering, and it is not 100 percent clear how this will interact with all the various differentiable rendering methods). If you do not mind, I would like to move forward with this at least for the command line frontend.